include/wx/valtext.h

```diff
@@ -63,43 +65,160 @@ class WXDLLIMPEXP_CORE wxTextValidator: public wxValidator
     void OnChar(wxKeyEvent& event);
 
     // ACCESSORS
-    inline long GetStyle() const { return m_validatorStyle; }
-    void SetStyle(long style);
+    inline long GetStyle() const { return m_style; }
+    void SetStyle(long style){ DoSetStyle(style); }
 
     wxTextEntry *GetTextEntry();
 
+    // strings & chars inclusions:
+    // ---------------------------
+
     void SetCharIncludes(const wxString& chars);
-    void SetIncludes(const wxArrayString& includes) { m_includes = includes; }
-    inline wxArrayString& GetIncludes() { return m_includes; }
+    void AddCharIncludes(const wxString& chars);
+
+    void SetIncludes(const wxArrayString& includes){ DoSetIncludes(includes); }
+    void AddInclude(const wxString& include){ DoAddInclude(include); }
+    const wxArrayString& GetIncludes() const { return m_includes; }
+    
+    wxString GetCharIncludes() const 
+    { 
+        if ( HasFlag(wxFILTER_INCLUDE_CHAR_LIST) )
+            return m_includes.Last();
```

Comment:
Agreed, it's best to introduce new members for char includes/excludes than the hackish approach used.

Reply:
Just remembered that adding new members for char includes/excludes makes those flags wxFILTER_{INCLUDE,EXCLUDE}_CHAR_LIST and wxFILTER_{INCLUDE,EXCLUDE}_LIST **useless**!

Reply:
Sorry, what do you mean? AFAICS in theory both pairs of flags make sense. I don't know about the implementation details however.

Reply:
I think those flags were used or introduced to discriminate `m_includes` and `m_excludes` ( because each was used for two different things, i.e. either hold an array of strings or an array of chars).

Now that the lists are separated, see [last commit](https://github.com/wxWidgets/wxWidgets/pull/1093/commits/d296ea8b3fb240a64ef3e59e53425addc0501762), we can leave those flags out and everything go without them.

Reply:
I could be completely missing something, but for me `CHAR` and non-`CHAR` versions just *mean* different things. `INCLUDE_CHAR`s means that only the following characters are allowed (e.g. only digits, or only letters). `INCLUDE` means that only the following _strings_ are allowed (e.g. only names of the rainbow colours), which is admittedly not very useful. `EXCLUDE_CHAR` means that the specified characters can not be used, which is again admittedly not very useful, but `EXCLUDE` means that the following strings are not accepted, which could be useful, I think.

In any case, they're just completely different, aren't they?

Reply:
Exactly what you just described. and IMSHO introducing {`INCLUDE`,`EXCLUDE`} `CHAR` and non-`CHAR` pairs in the design of `wxTextValidator` was a **mistake** because:
- they just complicated the design without much benefit.
- they prevented us from creating our validators in a natural way! take this:
        a textctrl that accepts only: **alphanumeric** chars and some other useful chars
```
// wxFILTER_INCLUDE_CHAR_LIST is not needed any longer with this PR
wxTextValidator val(wxFILTER_ALPHANUMERIC | wxFILTER_INCLUDE_CHAR_LIST);
val.SetCharIncludes(" ,'-");
```
the above snippet won't work in old implementation as one expects!

Reply:
One other observation:

transforming a list of chars - which are better represented as a one string - to an array of strings is not ideal neither. look at this:
```
// accepts digits only (plus and minus signs are ok!)
// wxFILTER_INCLUDE_CHAR_LIST is not needed any longer with this PR
wxTextValidator val(wxFILTER_DIGITS | wxFILTER_INCLUDE_CHAR_LIST);
val.SetCharIncludes("+-");
```
this won't work neither as currently implemented!

Reply:
Don't get me wrong, i don't claim it's buggy, but rather has shortcomings that can't be addressed without breaking things, unfortunately.

here is some of them:

1) cannot use `CHAR_LIST`s with its counterpart non-char `LIST`s together which may be useful for some validators! 
- so why cannot  we create validators that can **exclude** some strings and chars ? or is there any other way of creating such validators ?

2) I don't get why `wxFILTER_INCLUDE_LIST` is added at all ? because a textctrl that do not accept anything but a certain predefined strings, is hard to work with than a `wxCoice` or somthing similar.

3) why we have to split a `CHAR_LIST`s to an array of `wxString`s ?
is there any benefit doing so concerning space and performance or yet character encoding ?

4) why we can not use `wxFILTER_INCLUDE_CHAR_LIST` with one of these: `wxFILTER_ASCII`, `wxFILTER_ALPHA`, `wxFILTER_ALPHANUMERIC`, `wxFILTER_DIGITS`, `wxFILTER_NUMERIC` ?
- say i want a validator that accept **alpha** chars and some **ascii** (_apostrophe_, _comma_ and _hyphen_) only, how do you create such validator with the current implementation? 
this won't work :
```
wxTextValidator val(wxFILTER_ALPHA | wxFILTER_INCLUDE_CHAR_LIST);
val.SetCharIncludes("',-");
```
Thank you for your patience and help.

Reply:
(1) is a bug/limitation and it's nice that your changes fix it. (2) is indeed not particularly useful, but I don't see any real reason to remove it neither. (3) it's an implementation detail and it doesn't really matter much one way or the other. (4) is also a limitation and, again, it's good if it's fixed.

The only thing I want to avoid doing is breaking the existing code. These classes are old and there might be some weird uses of them in the existing code, so it's better not to change things unless really necessary.

Reply:
I anderstand your point. but what about adding a wxRegexTextValidator (or simply wxRegexValidator) as an enhanced version of wxTextValidator (with the same interface) but should not inherete from it ?

Reply:
Why not, I'm sure such class could be useful.

I'm not that enthusiastic about adding it right now because I still think we need to change the validators more in depth than that. As I mentioned in our `wxGenericValidator` discussion, I'd like to address many other problems with them and add support for

- Customizable error reporting, including highlighting of invalid input out of the box and showing validator-defined error messages in a program-controlled way.
- Extensible support for custom types.
- Ability to compose them.

etc.